cmd/snap/inhibit.go

```diff
@@ -34,143 +34,153 @@ import (
-		return err
+func (tf *textFlow) Finish() error {
+	if tf.pb != nil {
+		tf.pb.Finished()
 	}
 	return nil
 }
 
-var tryNotifyRefreshViaSnapDesktopIntegrationFlow = func(snapName string) (bool, error) {
+type graphicalFlow struct {
+	snapName string
+
+	notifiedDesktopIntegration bool
+}
+
+func (gf *graphicalFlow) Notify() error {
+	gf.notifiedDesktopIntegration = tryNotifyRefreshViaSnapDesktopIntegrationFlow(gf.snapName)
+	if gf.notifiedDesktopIntegration {
+		return nil
+	}
+
+	// unable to use snapd-desktop-integration, let's fallback graphical session flow
+	refreshInfo := client.PendingSnapRefreshInfo{
+		InstanceName: gf.snapName,
+		// remaining time = 0 results in "Snap .. is refreshing now" message from
+		// usersession agent.
+		TimeRemaining: 0,
+	}
+	return pendingRefreshNotification(&refreshInfo)
+}
+
+func (gf *graphicalFlow) Finish() error {
+	if gf.notifiedDesktopIntegration {
+		// snapd-desktop-integration detects inhibit unlock itself, do nothing
+		return nil
+	}
+
+	// finish graphical session flow
+	finishRefreshInfo := client.FinishedSnapRefreshInfo{InstanceName: gf.snapName}
+	return finishRefreshNotification(&finishRefreshInfo)
+}
+
+var tryNotifyRefreshViaSnapDesktopIntegrationFlow = func(snapName string) (notified bool) {
 	// Check if Snapd-Desktop-Integration is available
 	conn, err := dbusutil.SessionBus()
 	if err != nil {
 		logger.Noticef("unable to connect dbus session: %v", err)
-		return false, nil
+		return false
 	}
 	obj := conn.Object("io.snapcraft.SnapDesktopIntegration", "/io/snapcraft/SnapDesktopIntegration")
 	extraParams := make(map[string]dbus.Variant)
 	err = obj.Call("io.snapcraft.SnapDesktopIntegration.ApplicationIsBeingRefreshed", 0, snapName, runinhibit.HintFile(snapName), extraParams).Store()
 	if err != nil {
 		logger.Noticef("unable to successfully call io.snapcraft.SnapDesktopIntegration.ApplicationIsBeingRefreshed: %v", err)
-		return false, nil
+		return false
 	}
-	if _, err := waitInhibitUnlock(snapName, runinhibit.HintNotInhibited); err != nil {
-		return false, err
-	}
-	return true, nil
+	return true
 }
 
-func graphicalSessionFlow(snapName string, hint runinhibit.Hint) error {
-	refreshInfo := client.PendingSnapRefreshInfo{
-		InstanceName: snapName,
-		// Remaining time = 0 results in "Snap .. is refreshing now" message from
-		// usersession agent.
-		TimeRemaining: 0,
-	}
+var isGraphicalSession = func() bool {
+	// TODO: uncomment once there is a proper UX review
+	//return os.Getenv("DISPLAY") != "" || os.Getenv("WAYLAND_DISPLAY") != ""
+	return false
+}
 
-	if err := pendingRefreshNotification(&refreshInfo); err != nil {
-		return err
-	}
-	if _, err := waitInhibitUnlock(snapName, runinhibit.HintNotInhibited); err != nil {
+var pendingRefreshNotification = func(refreshInfo *client.PendingSnapRefreshInfo) error {
+	userclient := client.NewForUids(os.Getuid())
+	if err := userclient.PendingRefreshNotification(context.TODO(), refreshInfo); err != nil {
```

Comment:
Should context be a part of the `inhibitionFlow` interface?

Reply:
Basically anytime you use context.TODO it's a sign you have a problem in the call chain. Using context.Background is one way out but usually that's not the case.

What happens when `snap run` is interrupted while it has started doing `Pending RefreshNotifiction`? How does it cancel it?